File: src/main/java/net/dashflight/data/jwt/FingerprintService.java
package net.dashflight.data.jwt;

import com.google.common.hash.Hashing;
import java.nio.charset.StandardCharsets;
import java.security.SecureRandom;
import javax.xml.bind.DatatypeConverter;
import net.dashflight.data.random.LavaRandom;

/**
 * Utility for generating token fingerprints to strengthen security of JWTs
 */
public class FingerprintService {

    private static final int FINGERPRINT_LENGTH = 64;

    private SecureRandom secureRandom;

    public FingerprintService() {
        this.secureRandom = new LavaRandom();
    }

    /**
     * Generates a random fingerprint of length FINGERPRINT_LENGTH
     */
    public String generateRandomFingerprint() {
        byte[] randomBytes = new byte[FINGERPRINT_LENGTH];
        secureRandom.nextBytes(randomBytes);

        return DatatypeConverter.printHexBinary(randomBytes);
    }

    /**
     * Hashes a fingerprint with SHA-256
     */
    public String hashFingerprint(String fgp) {
        if (fgp == null) {
            return null;
        }

        byte[] fingerprintDigest = Hashing.sha256().hashBytes(fgp.getBytes(StandardCharsets.UTF_8)).asBytes();

        return DatatypeConverter.printHexBinary(fingerprintDigest);
    }

}

File: src/main/java/net/dashflight/data/jwt/FingerprintService.kt
package net.dashflight.data.jwt

import com.google.common.hash.Hashing
import com.google.inject.Inject
import java.nio.charset.StandardCharsets
import java.util.*
import javax.xml.bind.DatatypeConverter

/**
 * Utility for generating token fingerprints to strengthen security of JWTs
 */
class FingerprintService @Inject constructor(private val random: Random) {

    /**
     * Generates a random fingerprint of length FINGERPRINT_LENGTH
     */
    fun generateRandomFingerprint(): String {
        val randomBytes = ByteArray(FINGERPRINT_LENGTH)
        random.nextBytes(randomBytes)
        return DatatypeConverter.printHexBinary(randomBytes)
    }

    /**
     * Hashes a fingerprint with SHA-256
     */
    fun hashFingerprint(fgp: String?): String? {
        if (fgp == null) {
            return null
        }
        val fingerprintDigest = Hashing.sha256().hashBytes(fgp.toByteArray(StandardCharsets.UTF_8)).asBytes()
        return DatatypeConverter.printHexBinary(fingerprintDigest)
    }

    companion object {
        private const val FINGERPRINT_LENGTH = 64
    }

}